File: swift_agent/Sources/main/computers/macos/macos.swift

import AppKit
import Foundation
import CoreGraphics
import Carbon

class MacOSComputer: Computer {
    let initialUrl: String
    let highlightMouse: Bool
    var currentUrl: String = "macos://desktop"

    let MACOS_KEY_MAP: [String: CGKeyCode] = [
        "control": CGKeyCode(kVK_Control),
        "ctrl": CGKeyCode(kVK_Control),
        "shift": CGKeyCode(kVK_Shift),
        "alt": CGKeyCode(kVK_Option),
        "option": CGKeyCode(kVK_Option),
        "command": CGKeyCode(kVK_Command),
        "cmd": CGKeyCode(kVK_Command),
        "meta": CGKeyCode(kVK_Command),
        "enter": CGKeyCode(kVK_Return),
        "return": CGKeyCode(kVK_Return),
        "tab": CGKeyCode(kVK_Tab),
        "escape": CGKeyCode(kVK_Escape),
        "esc": CGKeyCode(kVK_Escape),
        "space": CGKeyCode(kVK_Space),
        "pageup": CGKeyCode(kVK_PageUp),
        "pagedown": CGKeyCode(kVK_PageDown),
        "f1": CGKeyCode(kVK_F1),
        "f2": CGKeyCode(kVK_F2),
        "f3": CGKeyCode(kVK_F3),
        "f4": CGKeyCode(kVK_F4),
        "f5": CGKeyCode(kVK_F5),
        "f6": CGKeyCode(kVK_F6),
        "f7": CGKeyCode(kVK_F7),
        "f8": CGKeyCode(kVK_F8),
        "f9": CGKeyCode(kVK_F9),
        "f10": CGKeyCode(kVK_F10),
        "f11": CGKeyCode(kVK_F11),
        "f12": CGKeyCode(kVK_F12),
        "left": CGKeyCode(kVK_LeftArrow),
        "right": CGKeyCode(kVK_RightArrow),
        "up": CGKeyCode(kVK_UpArrow),
        "down": CGKeyCode(kVK_DownArrow),
        "delete": CGKeyCode(kVK_ForwardDelete),
        "backspace": CGKeyCode(kVK_Delete)
    ]

    init(initialUrl: String, highlightMouse: Bool) {
        self.initialUrl = initialUrl
        self.highlightMouse = highlightMouse
    }

    func screenSize() -> (Int, Int) {
        if let screen = NSScreen.main {
            let rect = screen.frame
            return (Int(rect.size.width), Int(rect.size.height))
        }
        return (0, 0)
    }

    func clickAt(x: Int, y: Int) -> EnvState {
        moveMouse(x: x, y: y)
        let mouseLocation = CGPoint(x: x, y: y)
        let mouseDown = CGEvent(mouseEventSource: nil, mouseType: .leftMouseDown, mouseCursorPosition: mouseLocation, mouseButton: .left)
        let mouseUp = CGEvent(mouseEventSource: nil, mouseType: .leftMouseUp, mouseCursorPosition: mouseLocation, mouseButton: .left)
        mouseDown?.post(tap: .cghidEventTap)
        mouseUp?.post(tap: .cghidEventTap)
        waitForUI()
        return currentState()
    }

    func hoverAt(x: Int, y: Int) -> EnvState {
        moveMouse(x: x, y: y)
        waitForUI()
        return currentState()
    }

    func typeTextAt(x: Int, y: Int, text: String, pressEnter: Bool, clearBeforeTyping: Bool) -> EnvState {
        moveMouse(x: x, y: y)
        clickAt(x: x, y: y)
        waitForUI()

        if clearBeforeTyping {
            keyCombination(keys: ["command", "a"])
            keyCombination(keys: ["delete"])
            waitForUI()
        }

        for char in text {
            typeKey(key: String(char))
            waitForUI(delay: 0.01)
        }

        if pressEnter {
            keyCombination(keys: ["enter"])
            waitForUI()
        }

        return currentState()
    }

    func scrollDocument(direction: String) -> EnvState {
        let (width, height) = screenSize()
        return scrollAt(x: width / 2, y: height / 2, direction: direction, magnitude: 600)
    }

    func scrollAt(x: Int, y: Int, direction: String, magnitude: Int) -> EnvState {
        moveMouse(x: x, y: y)
        let scrollEvent = CGEvent(scrollWheelEvent2Source: nil, units: .pixel, wheelCount: 1, wheel1: Int32(direction == "down" ? -magnitude : (direction == "up" ? magnitude : 0)), wheel2: Int32(direction == "right" ? -magnitude : (direction == "left" ? magnitude : 0)), wheel3: 0)
        scrollEvent?.post(tap: .cghidEventTap)
        waitForUI()
        return currentState()
    }

    func wait5Seconds() -> EnvState {
        Thread.sleep(forTimeInterval: 5)
        return currentState()
    }

    func goBack() -> EnvState {
        keyCombination(keys: ["command", "["])
        waitForUI()
        return currentState()
    }

    func goForward() -> EnvState {
        keyCombination(keys: ["command", "]"])
        waitForUI()
        return currentState()
    }

    func search() -> EnvState {
        return navigate(url: "https://www.google.com")
    }

    func navigate(url: String) -> EnvState {
        var normalizedUrl = url
        if !normalizedUrl.hasPrefix("http://") && !normalizedUrl.hasPrefix("https://") {
            normalizedUrl = "https://" + normalizedUrl
        }
        let process = Process()
        process.launchPath = "/usr/bin/open"
        process.arguments = [normalizedUrl]
        process.launch()
        currentUrl = normalizedUrl
        Thread.sleep(forTimeInterval: 2)
        return currentState()
    }

    func keyCombination(keys: [String]) -> EnvState {
        let keyCodes = keys.map { MACOS_KEY_MAP[$0.lowercased()] ?? 0 }
        let source = CGEventSource(stateID: .hidSystemState)

        for keyCode in keyCodes {
            let keyDown = CGEvent(keyboardEventSource: source, virtualKey: keyCode, keyDown: true)
            keyDown?.post(tap: .cghidEventTap)
        }

        for keyCode in keyCodes.reversed() {
            let keyUp = CGEvent(keyboardEventSource: source, virtualKey: keyCode, keyDown: false)
            keyUp?.post(tap: .cghidEventTap)
        }

        waitForUI()
        return currentState()
    }

    func dragAndDrop(x: Int, y: Int, destinationX: Int, destinationY: Int) -> EnvState {
        moveMouse(x: x, y: y)
        let mouseDown = CGEvent(mouseEventSource: nil, mouseType: .leftMouseDown, mouseCursorPosition: CGPoint(x: x, y: y), mouseButton: .left)
        mouseDown?.post(tap: .cghidEventTap)
        waitForUI()
        moveMouse(x: destinationX, y: destinationY)
        let mouseUp = CGEvent(mouseEventSource: nil, mouseType: .leftMouseUp, mouseCursorPosition: CGPoint(x: destinationX, y: destinationY), mouseButton: .left)
        mouseUp?.post(tap: .cghidEventTap)
        waitForUI()
        return currentState()
    }

    func currentState() -> EnvState {
        let displayID = CGMainDisplayID()
        let image = CGDisplayCreateImage(displayID)
        let bitmap = NSBitmapImageRep(cgImage: image!)
        let data = bitmap.representation(using: .png, properties: [:])!
        return EnvState(screenshot: data, url: currentUrl)
    }

    func close() {}

    private func moveMouse(x: Int, y: Int) {
        let mouseLocation = CGPoint(x: x, y: y)
        let mouseMove = CGEvent(mouseEventSource: nil, mouseType: .mouseMoved, mouseCursorPosition: mouseLocation, mouseButton: .left)
        mouseMove?.post(tap: .cghidEventTap)
    }

    private func waitForUI(delay: TimeInterval = 0.5) {
        Thread.sleep(forTimeInterval: delay)
    }

    private func typeKey(key: String) {
        let source = CGEventSource(stateID: .hidSystemState)
        let keyDown = CGEvent(keyboardEventSource: source, virtualKey: 0, keyDown: true)
        keyDown?.keyboardSetUnicodeString(stringLength: key.count, unicodeString: [unichar](key.utf16))
        keyDown?.post(tap: .cghidEventTap)

        let keyUp = CGEvent(keyboardEventSource: source, virtualKey: 0, keyDown: false)
        keyUp?.keyboardSetUnicodeString(stringLength: key.count, unicodeString: [unichar](key.utf16))
        keyUp?.post(tap: .cghidEventTap)
    }
}

File: computers/macos/macos.py
import io
import subprocess
import time
from typing import Literal, Optional

import pyautogui

from ..computer import Computer, EnvState


MACOS_KEY_MAP = {
    "control": "ctrl",
    "ctrl": "ctrl",
    "shift": "shift",
    "alt": "option",
    "option": "option",
    "command": "command",
    "cmd": "command",
    "meta": "command",
    "enter": "enter",
    "return": "enter",
    "tab": "tab",
    "escape": "esc",
    "esc": "esc",
    "space": "space",
    "pageup": "pageup",
    "pagedown": "pagedown",
    "f1": "f1",
    "f2": "f2",
    "f3": "f3",
    "f4": "f4",
    "f5": "f5",
    "f6": "f6",
    "f7": "f7",
    "f8": "f8",
    "f9": "f9",
    "f10": "f10",
    "f11": "f11",
    "f12": "f12",
    "left": "left",
    "right": "right",
    "up": "up",
    "down": "down",
    "delete": "delete",
    "backspace": "backspace",
}


def _normalize_key(key: str) -> str:
    return MACOS_KEY_MAP.get(key.lower(), key.lower())


class MacOSComputer(Computer):
    """Computer implementation that sends real macOS input events."""

    def __init__(
        self,
        screen_size: Optional[tuple[int, int]] = None,
        initial_url: str = "https://www.google.com",
        search_engine_url: str = "https://www.google.com",
        highlight_mouse: bool = False,
    ):
        # Setting FAILSAFE to False prevents pyautogui from raising if the cursor
        # hits a screen corner while the agent moves the mouse.
        pyautogui.FAILSAFE = False
        self._screen_size_override = screen_size
        self._initial_url = initial_url
        self._search_engine_url = search_engine_url
        self._highlight_mouse = highlight_mouse
        self._current_url = "macos://desktop"

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        return False

    def screen_size(self) -> tuple[int, int]:
        if self._screen_size_override:
            return self._screen_size_override
        size = pyautogui.size()
        return size.width, size.height

    def open_web_browser(self) -> EnvState:
        # Open the initial URL in the default browser.
        self.navigate(self._initial_url)
        return self.current_state()

    def click_at(self, x: int, y: int) -> EnvState:
        self._move_mouse(x, y)
        pyautogui.click(x=x, y=y)
        self._wait_for_ui()
        return self.current_state()

    def hover_at(self, x: int, y: int) -> EnvState:
        self._move_mouse(x, y)
        self._wait_for_ui()
        return self.current_state()

    def type_text_at(
        self,
        x: int,
        y: int,
        text: str,
        press_enter: bool,
        clear_before_typing: bool,
    ) -> EnvState:
        self._move_mouse(x, y)
        pyautogui.click(x=x, y=y)
        self._wait_for_ui()

        if clear_before_typing:
            pyautogui.hotkey("command", "a")
            pyautogui.press("delete")
            self._wait_for_ui()

        pyautogui.typewrite(text, interval=0.01)
        self._wait_for_ui()

        if press_enter:
            pyautogui.press("enter")
            self._wait_for_ui()

        return self.current_state()

    def scroll_document(
        self, direction: Literal["up", "down", "left", "right"]
    ) -> EnvState:
        center_x, center_y = (value // 2 for value in self.screen_size())
        return self.scroll_at(center_x, center_y, direction, magnitude=600)

    def scroll_at(
        self,
        x: int,
        y: int,
        direction: Literal["up", "down", "left", "right"],
        magnitude: int,
    ) -> EnvState:
        self._move_mouse(x, y)
        if direction == "up":
            pyautogui.scroll(magnitude)
        elif direction == "down":
            pyautogui.scroll(-magnitude)
        elif direction == "left":
            pyautogui.hscroll(magnitude)
        elif direction == "right":
            pyautogui.hscroll(-magnitude)
        else:
            raise ValueError(f"Unsupported direction: {direction}")
        self._wait_for_ui()
        return self.current_state()

    def wait_5_seconds(self) -> EnvState:
        time.sleep(5)
        return self.current_state()

    def go_back(self) -> EnvState:
        pyautogui.hotkey("command", "[")
        self._wait_for_ui()
        return self.current_state()

    def go_forward(self) -> EnvState:
        pyautogui.hotkey("command", "]")
        self._wait_for_ui()
        return self.current_state()

    def search(self) -> EnvState:
        return self.navigate(self._search_engine_url)

    def navigate(self, url: str) -> EnvState:
        normalized_url = url
        if not normalized_url.startswith(("http://", "https://")):
            normalized_url = "https://" + normalized_url
        subprocess.Popen(["open", normalized_url])
        self._current_url = normalized_url
        # Give the browser time to open or switch tabs.
        time.sleep(2)
        return self.current_state()

    def key_combination(self, keys: list[str]) -> EnvState:
        normalized_keys = [_normalize_key(k) for k in keys]
        pyautogui.hotkey(*normalized_keys)
        self._wait_for_ui()
        return self.current_state()

    def drag_and_drop(
        self, x: int, y: int, destination_x: int, destination_y: int
    ) -> EnvState:
        self._move_mouse(x, y)
        pyautogui.mouseDown(x=x, y=y)
        self._wait_for_ui()
        self._move_mouse(destination_x, destination_y)
        pyautogui.mouseUp(x=destination_x, y=destination_y)
        self._wait_for_ui()
        return self.current_state()

    def current_state(self) -> EnvState:
        screenshot = pyautogui.screenshot()
        buffer = io.BytesIO()
        screenshot.save(buffer, format="PNG")
        return EnvState(screenshot=buffer.getvalue(), url=self._current_url)

    def _move_mouse(self, x: int, y: int):
        pyautogui.moveTo(x, y, duration=0.15 if self._highlight_mouse else 0)

    def _wait_for_ui(self, delay: float = 0.5):
        time.sleep(delay)
